### PR TITLE
feat(slider.tsx): implement slider component

### PR DESCRIPTION
**Pull request by migue1coel6odev**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: modules/slider/Slider.tsx [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    "&::-moz-range-thumb": {},
+    "&::-ms-thumb": {},
+    "&:focus": {},
+    "&::-ms-track": {},
+    "&::-webkit-slider-runnable-track": {}
+  },
+  value: {}
+});
+
+const calculateLeftSpace = (
```
Comment (diogo-mendes) [inline modules/slider/Slider.tsx]: If you define this function inside a `useCallback` in the component, you can simplify the function signature by removing most parameters 💭 although I do understand why you defined it outside! Just a consideration

File: modules/slider/Slider.tsx [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+const Slider: React.FC<ISliderProps & WithStylesProps> = ({
+  cx,
+  styles,
+  min,
+  max,
+  value,
+  step,
+  onChange,
+  disabled = false,
+  className = ""
```
Comment (diogo-mendes) [inline modules/slider/Slider.tsx]: Is the default value necessary? Also, I'm not sure if placing the classname at the `input` level and not the `wrapper` level is the best idea, simply because I see it being used for setting margins and padding and whatnot. With the current implementation, that won't be possible.

File: modules/slider/Slider.tsx (hunk 1 above)
Comment (diogo-mendes) [reply modules/slider/Slider.tsx]: Also, this formula is a bit hard to understand, at least at first glance. Do you think adding a comment describing what and how this function works would be helpful, or do you believe it is already as self-descriptive as possible?

File: modules/slider/Slider.story.mdx [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+## Disabled Slider
+
+<Preview>
+  <Story name="Disabled Slider">
+    <DisabledSliderStory />
+  </Story>
+</Preview>
+
+## Locked Slider
```
Comment (diogo-mendes) [inline modules/slider/Slider.story.mdx]: Don't forget to add the available props and extendable styles for both `Slider` and `LockedSlider` 🙂

File: modules/slider/Slider.tsx [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      </div>
+      <input
+        ref={ref}
+        className={cx(styles.input, disabled && "disabled", className)}
+        type="range"
+        min={min}
+        max={max}
+        value={value}
+        step={step}
+        onChange={ev => onChange?.(ev.currentTarget.value)}
```
Comment (diogo-mendes) [inline modules/slider/Slider.tsx]: If `onChange` is a mandatory prop, do you need to check if it's there?

File: modules/slider/Slider.tsx [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+import { withStyles } from "@diana-ui/base";
+import { ThemeStyleSheetFactory, WithStylesProps } from "@diana-ui/types";
+import { LabelRegular } from "@diana-ui/typography";
+
+export interface ISliderProps {
+  min: number;
+  max: number;
+  value: number;
+  step: number;
+  onChange: any;
```
Comment (diogo-mendes) [inline modules/slider/Slider.tsx]: `onChange: (value: number) => void;` ?

File: modules/slider/Slider.tsx [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  disabled?: boolean;
+  className?: string;
+}
+
+export interface ISliderState {
```
Comment (diogo-mendes) [inline modules/slider/Slider.tsx]: Is this being used?

File: modules/slider/Slider.tsx [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+          min,
+          max,
+          value
+        )
+      );
+    }
+
+    updateLeftSpacing();
+
+    window.addEventListener("resize", updateLeftSpacing);
```
Comment (diogo-mendes) [inline modules/slider/Slider.tsx]: We already have a hook for window resizes. I think you could use it here, what do you think? It would make the code here simpler IMO 🙂

File: modules/slider/Slider.tsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  disabled = false,
+  className = ""
+}) => {
+  const [size, setSize] = useState(9);
+  const ref = React.createRef<HTMLInputElement>();
+  useEffect(() => {
+    function updateLeftSpacing() {
+      setSize(
+        calculateLeftSpace(
+          ref.current?.clientWidth ? ref.current.clientWidth - 21 : 0,
```
Comment (diogo-mendes) [inline modules/slider/Slider.tsx]: There are a lot of magic numbers overall in this file. Can we assign them to constant variables or at least describe what they represent with comments? It's a bit hard to follow what is going on otherwise

File: modules/slider/LockedSlider.tsx [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+const styleSheet: ThemeStyleSheetFactory = theme => ({
+  lockWrapper: {
+    display: "flex",
+    flexDirection: "row",
+    alignItems: "flex-end",
+    width: "100%"
+  },
+  icon: {
+    marginLeft: 16
```
Comment (diogo-mendes) [inline modules/slider/LockedSlider.tsx]: Use `theme.spaceUnit`?

File: modules/slider/LockedSlider.tsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  styles,
+  onChange,
+  startLocked = false,
+  lockedIcon,
+  unlockedIcon,
+  ...props
+}) => {
+  const [isLocked, setIsLocked] = useState(startLocked);
+  return (
+    <div className={cx(styles.lockWrapper)}>
```
Comment (diogo-mendes) [inline modules/slider/LockedSlider.tsx]: We should allow a `className` prop to be set here. We just need to see how that works with `Slider`'s className logic. Maybe pass `className=""` down to `Slider` here?

File: modules/slider/LockedSlider.tsx [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+      <StyledSlider
+        className={cx(isLocked && "locked")}
+        onChange={(!isLocked && onChange) || (() => {})}
```
Comment (diogo-mendes) [inline modules/slider/LockedSlider.tsx]: Why do we need the fallback if onChange is mandatory?

File: modules/slider/LockedSlider.tsx [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        {...props}
+      />
+      <div className={cx(styles.icon)} onClick={() => setIsLocked(!isLocked)}>
+        {isLocked &&
+          (typeof lockedIcon === "string" ? (
+            <Icon className={cx(styles.icon)} name={lockedIcon as any} />
```
Comment (diogo-mendes) [inline modules/slider/LockedSlider.tsx]: We are applying the same `styles.icon` to the icon wrapper and the icon itself? 🤔

File: modules/slider/LockedSlider.tsx [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+          ) : (
+            lockedIcon
+          ))}
+        {!isLocked &&
```
Comment (diogo-mendes) [inline modules/slider/LockedSlider.tsx]: I think we could simplify this by defining 
`const icon = isLocked ? lockedIcon : unlockedIcon;`
outside the render, and here just have 
```
<div className={cx(styles.icon)} onClick={() => setIsLocked(!isLocked)}>
  (typeof icon === "string" ? (
    <Icon className={cx(styles.icon)} name={icon as any} />
  ) : (
    icon
  ))
</div>
```

What do you think?

File: modules/slider/LockedSlider.tsx [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    flexDirection: "row",
+    alignItems: "flex-end",
+    width: "100%"
+  },
+  icon: {
+    marginLeft: 16
+  }
+});
+
+// const LOCKED_THUMB =
```
Comment (diogo-mendes) [inline modules/slider/LockedSlider.tsx]: Is this supposed to be here? Can it be removed?

File: modules/slider/LockedSlider.tsx (hunk 11 above)
Comment (nunoliveira12) [reply modules/slider/LockedSlider.tsx]: Maybe it's better to separate the logic as a callback function:

const onChangeCallback = useCallback ((e) => {
   if(!isLocked) {
      onChange(e)
}, [isLocked, onChange]);

File: modules/slider/LockedSlider.tsx [hunk 15]
```diff
@@ -0,0 +1,65 @@
+import React, { useState } from "react";
+import { withStyles } from "@diana-ui/base";
+import { ThemeStyleSheetFactory, WithStylesProps } from "@diana-ui/types";
+import { Icon } from "@diana-ui/icon";
+import Slider, { ISliderProps } from "./Slider";
+
+export interface ILockedSliderProps extends ISliderProps {
+  startLocked?: boolean;
+  lockedIcon: string | React.ReactNode;
```
Comment (nunoliveira12) [inline modules/slider/LockedSlider.tsx]: Could we just choose one way of having it? seems confusing to have for example lockedIcon as string and unlockedIcon as Icon element

File: modules/slider/README.md [hunk 16]
```diff
@@ -0,0 +1 @@
+## @diana-ui/slider
```
Comment (nunoliveira12) [inline modules/slider/README.md]: As @diogo-mendes likes to remind 😆 , write here some description of component. Something small

File: modules/slider/Slider.tsx [hunk 17]
```diff
@@ -0,0 +1,101 @@
+import React, { useState, useEffect } from "react";
+import { withStyles } from "@diana-ui/base";
+import { ThemeStyleSheetFactory, WithStylesProps } from "@diana-ui/types";
+import { LabelRegular } from "@diana-ui/typography";
+
+export interface ISliderProps {
```
Comment (nunoliveira12) [inline modules/slider/Slider.tsx]: you should extend StandardProps<"input"> and then pass them to the input. This way we allow some standard input props that we can later be able to use, such as aria-label, classname,( disable, min, max, value, step ), etc

File: modules/slider/Slider.tsx (hunk 1 above)
Comment (nunoliveira12) [reply modules/slider/Slider.tsx]: I would also add it inside as a useCallback. It would be easier to understand and it would remove params complexity

File: modules/slider/Slider.tsx [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+  max: number,
+  value: number
+) => (value * totalWidth) / (max - min) + 13 - (`${max}`.length * 20) / 2;
+
+const Slider: React.FC<ISliderProps & WithStylesProps> = ({
+  cx,
+  styles,
+  min,
+  max,
+  value,
```
Comment (nunoliveira12) [inline modules/slider/Slider.tsx]: My opinion here, I prefer to let this components have their own controlled state (or as we call it, to be uncontrolled). 
This means, if you dont pass value, have something like:

const [_value, setValue] = useState(initialValue);

So it would always have its own value. This way we can move the slider and fire onChanges (if needed) without the need of the parent to pass the Slider the value itself. 

Of course we must keep the option of the value to be passed to the component as it is.
Usually as:

useEffect(() => {
if(value !== undefined)
setValue(value)
})

this way if it has a value, it becomes controlled by their parent.

File: modules/tokens/themes/theme.default.types.ts [hunk 19]
```diff
@@ -1,6 +1,7 @@
 export enum FontWeight {
-    REGULAR = 400,
-    MEDIUM = 500,
-    BOLD = 700,
-    BOLDER = 900,
+  REGULAR = 400,
+  MEDIUM = 500,
+  BOLD = 700,
+  BOLDER = 900,
+  NORMAL = "normal"
```
Comment (nunoliveira12) [inline modules/tokens/themes/theme.default.types.ts]: the normal here is 400 as REGULAR, you sure it is needed to have this option?

File: modules/slider/Slider.tsx [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+  onValueChange,
+  disabled = false,
+  className = "",
+  inputClassName = ""
+}) => {
+  const windowSize = useWindowSize();
+  const [size, setSize] = useState(9);
+  const [_value, setValue] = useState(value || 0);
+  const ref = React.createRef<HTMLInputElement>();
+
```
Comment (nunoliveira12) [inline modules/slider/Slider.tsx]: you might need to force setValue(value) on an useEffect, otherwise if the value is changed on the parent, it won't get replaced, right?

example:

//everytime the parent sets a new value, it will have priority over its own state
useEffect(() => {
  if(value !== undefined) {
    setValue(value); 
  }
}, [value])

File: modules/tokens/themes/theme.default.types.ts (hunk 19 above)
Comment (nunoliveira12) [reply modules/tokens/themes/theme.default.types.ts]: bump
